File: problem5.py
''' 
find smallest number which is evenly divisible by all numbers from 1 to 20

Create initial array from 2 to 20.

find prime numbers between 2 and 20. Add to primeFactorArray, remove form initial array

find prime factors of remaining numbers.

if prime factors already within array, ignore, if not add to primeFactorArray

multiply primeFactorArray numbers together to find answer.
'''

def primeNumber():
	
	for testNum in testArray:
		
		if testNum == 2:
			
			primeFactorArray.append(testNum)
		
		else:
			
			test = True
			j = 2
			while j < testNum:
				if testNum % j == 0:
					test = False
					j = testNum
				j = j + 1
				
			if test == True:
				primeFactorArray.append(testNum)
	
	for number in primeFactorArray:
		testArray.remove(number)
		
def findPrimeFactors(number):
		
	primeFactorsTemp = []
	divisor = 2
		
	while number > 1:
		while number % divisor == 0:
			primeFactorsTemp.append(divisor)
			number = number / divisor
	
		divisor = divisor + 1
	
	return primeFactorsTemp

# program starts here
primeFactorArray = []
testArray = []
prompt = int(raw_input("Choose a number: "))

for i in range(2, prompt + 1):
	testArray.append(i)

primeNumber()

for number in testArray:
	primeFactorsTemp = findPrimeFactors(number)

	primeFactorBuffer = []
	for number in primeFactorsTemp:
	
		if number in primeFactorArray:
			primeFactorBuffer.append(number)
			primeFactorArray.remove(number)
		else:
			primeFactorArray.append(number)
	
	for number in primeFactorBuffer:
		primeFactorArray.append(number)

answer = 1
for number in primeFactorArray:
	answer = answer * number

print "answer is: %d" % answer

File: problem5.js
/* 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20? */

/* must be multiple of 380 */

/* Find the smallest set of primes, such that all numbers 1-20 can be constructed. This set will 
be the prime factorisation of the smallest number to which 1-20 are all evenly divisible. */

var primeArray = [2,3,5,7,11,13,17,19];
var primeObject = {
    2: [],
    3: [],
    5: [],
    7: [],
    11: [],
    13: [],
    17: [],
    19: []
};
var start = 11;
var end = 20;
var answer = 1;

/* find prime factors */
var primeFactors = function(number) {
    factors = [];
	divisor = 2;
	var container = new Object(); 
	
	while ( number > 1 ) {
		while ( number % divisor === 0 ){
			factors.push(divisor);
			number = number / divisor ;
		}

		divisor++ ;
	}
    
    container.number = factors;
    
	return container;
};

for (i=start ; i <= end ; i++ ) {
	
	var container = primeFactors(i);
	
	/* compare container to existing object, add missing factors */
	for (var key in container){
	    pOLength = primeObject.key.length;
	    cLength = container.key.length;
	}
	
	
	for ( j = 0; j < pFArray.length ; j++  ) {
		primeArray.push(pFArray[j]);
	}
	
}

for ( i=0 ; i < primeArray.length ; i++ ){
	answer = answer * primeArray[i];
}
console.log(primeFactors);
console.log(answer);

